### DISCUSSION TITLE
How to Compile Istio for RISC-V Architecture？

### DISCUSSION THREAD

**Discussion by Boring545** [istio/istio#51764] [Q&A]
I would like to compile Istio for the RISC-V architecture. I am seeking guidance on whether this is feasible and how to generate RISC-V compatible versions of the images used in the build process.
Is it feasible to compile Istio for the RISC-V architecture?
What steps are required to generate RISC-V compatible versions of the images used in the build process?

**Comment by zirain** [comment]
I think it's be fine for golang components, the challenge should be ztunnel and proxy(aka envoy).

**Comment by ilrudie** [comment]
I don't have a risc system to try it on but super quick ztunnel test got pretty far with just the cross-compile toolchain, it failed on riscv64-linux-gnu-gcc missing which is a extra toolchain bit that seems to be needed for ring (tls). You might have some luck with zt after all. If you have rust installed on your system you could clone the ztunnel repo and run `cargo build` to see if it will even compile.

Interested in how it works for you. Please report back.

**Reply by Boring545** [reply]
```
   Compiling hickory-proto v0.24.1
    Building [=======================> ] 365/371: hickory-proto                                                                                         
   Compiling hickory-resolver v0.24.1
   Compiling hickory-client v0.24.1
   Compiling hickory-server v0.24.1
   Compiling ztunnel v0.0.0 (/home/zjq/build_factory/ztunnel/ztunnel)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 7m 4
```
I successfully compiled ztunnel in my RISC-V64 environment without encountering any issues. Now I am trying to compile Istio, and I noticed that a Docker image called build-tools is needed. Should I attempt to build this image manually?

**Reply by ilrudie** [reply]
It's probably the easiest way to satisfy all dependencies if it will build for you

**Comment by Boring545** [comment]
After some effort, I successfully built an old version of Envoy (1.26.4). The newer version of Envoy cannot be built at the moment due to some RISC-V compatibility issues. Based on this, I am trying to build the proxy. Is it correct to use the command `bazel build -c --fission=no opt envoy`?

**Reply by zirain** [reply]
`make build` under proxy repo

**Reply by Boring545** [reply]
Thank you! If I have any progress, I will provide feedback here.

**Comment by Boring545** [comment]
I successfully built proxy 1.18.2, which is a relatively older version. The reason I built it is that due to some Bazel-related issues, I was unable to build a newer version of Envoy. Currently, I have successfully built both ztunnel and proxy. Should I choose Istio 1.18.2 for building? In fact, I haven't found a way to integrate my proxy into Istio. I noticed that `envoy/bin/init.sh` downloads something from `https://storage.googleapis.com/istio-build`. Can you tell me what is being downloaded? Is it the proxy? How should I package my build results?

**Reply by zirain** [reply]
it won't download if there's a binary, what you need to do is copy you envoy to there.

**Reply by Boring545** [reply]
I will try to do that, thank you! Perhaps you could give me some advice on which Istio version I should choose for building?

**Reply by zirain** [reply]
as latest as you can

**Comment by Boring545** [comment]
Based on proxy 1.18.2 and ztunnel-1.23.0, I successfully built istio-1.22.2. The reason for not choosing the latest version is that I have a build-tools image already built for istio-1.22.2. After the successful build, I encountered an error when executing make test, specifically with the test item istio-agent. You can refer to the detailed error log [here](https://github.com/Boring545/bazeltest/blob/main/bugrecord/bug6.txt). Could you please provide me with some suggestions?

**Comment by Boring545** [comment]
### Problem Summary

I'm experiencing several errors related to the Istio and Envoy setup. Below are the key error types observed in the logs:

1. **Wasm Module Fetch Error**
   ```
   error   wasm    convert the wasm config: cannot fetch Wasm module https://test-url: error
   warn    ads     ADS:ECDS: ACK ERROR debug-86 Internal:cannot fetch Wasm module https://test-url: error
   ```
   - **Summary**: The system fails to fetch the Wasm module from the specified URL. This could be due to network issues, an incorrect URL, or the target service being unavailable.

2. **Missing Istio Version**
   ```
   warn    model   Istio Version is not found in metadata for debug, which may have undesirable side effects
   ```
   - **Summary**: The Istio version information is not found in the metadata, which may lead to undesirable side effects impacting the system's normal operation.

3. **Envoy Process Crash**
   ```
   error Envoy exited with error: signal: illegal instruction (core dumped)
   ```
   - **Summary**: The Envoy process crashed due to an illegal instruction. This could be caused by binary incompatibility, incorrect compilation options, or other system-level issues.

4. **Citadel Client Connection Issues**
   ```
   error citadelclient failed to sign CSR: create certificate: rpc error: code = Canceled desc = grpc: the client connection is closing
   error citadelclient failed reconnect: failed to close connection: rpc error: code = Canceled desc = grpc: the client connection is closing
   error cache resource:default failed to sign: create certificate: rpc error: code = Canceled desc = grpc: the client connection is closing
   ```
   - **Summary**: The Citadel client failed to sign the certificate signing request (CSR) because the gRPC client connection is closing. Multiple components encountered similar connection closure issues, which may relate to network instability or CA service failure.

5. **Authentication Failures**
   ```
   error   security        Failed to authenticate client from 127.0.0.1:60436: Authenticator fake: neither token (jwt authentication not allowed) nor cert (cert authentication not allowed) succeeded
   error   citadelclient   failed to sign CSR: create certificate: rpc error: code = Unauthenticated desc = request authenticate failure
   error   security        Failed to authenticate client from 127.0.0.1:50824: Authenticator fake: neither token (jwt authentication not allowed) nor cert (no verified chain is found) succeeded
   ```
   - **Summary**: Authentication failed due to the absence of a valid token (JWT authentication not allowed) or certificate (certificate authentication not allowed). This indicates potential misconfiguration in the authentication mechanism.

**Comment by tony-code25** [comment]
请问编译成功了吗？可以分享下经验吗

**Reply by Boring545** [reply]
You might want to check out https://github.com/Boring545/istio/tree/1.22.2-riscv — I ran into an issue with the agent during the final testing stage, but didn’t look into it any further.
